### DISCUSSION TITLE
Table or Row : cannot handle various images

### DISCUSSION THREAD

**Discussion by KaniSchotski** [QuestPDF/QuestPDF#447] [Q&A]
I want to put several images next to each other on the same row in a questPDF table or row.

It works with only 1 image. But as soon as I take several (even very small images) QuestPDF does no longer return a PDF (even no error message).

Example (imageBytes is an image of type byte[]) :

- table.Cell().Row(1).Column(1).Image(imageBytes, ImageScaling.FitArea);
- table.Cell().Row(1).Column(2).Image(imageBytes, ImageScaling.FitArea);
- ... up to 5 columns on a row

- row.RelativeItem().Image(imageBytes, ImageScaling.FitArea);
- row.RelativeItem().Image(imageBytes, ImageScaling.FitArea);
- ... up to 5 RelativeItems on a row

When replacing the image by PlaceHolders.Image in the table or row, everything works perfect 👍 

- table.Cell().Row(1).Column(1).Image(Placeholders.Image(100,100),ImageScaling.FitWidth);
- table.Cell().Row(1).Column(2).Image(Placeholders.Image(100,100),ImageScaling.FitWidth);

- row.RelativeItem().Image(Placeholders.Image(100, 100), ImageScaling.FitWidth);
- row.RelativeItem().Image(Placeholders.Image(100, 100), ImageScaling.FitWidth);

When adding the image as a last item to a row in a table or on a row, the image shows. But it does not show when inserting it in between the first and last item of a row.

It cannot be a problem with the image, because in some cases it shows up, in others it doesn't. Below images with corresponding code.

Please help. QuestPDF is a fantastic framework. I would love to use it more intensively for catalogue generation.

Thanks in advance !
Kani

![image](https://user-images.githubusercontent.com/94176942/210553975-bcc2b5e6-b5c6-43b1-803a-fbd265eb3295.png)

![image](https://user-images.githubusercontent.com/94176942/210554802-da6fb40e-0800-482e-be43-6454e9b5b23d.png)

![image](https://user-images.githubusercontent.com/94176942/210555608-984f1cd3-c99a-4089-af1b-ded4084d2ea9.png)

**Comment by MarcinZiabek** [comment]
Hi, would you be able to provide a minimal example of the problem? Best as repository? I would love to better understand this issue and fix it 😁

My gut tells me that this might be yet another issue with floating-point arithmetic. A very rare corner case. I have an idea how to fix it but it cannot be done overnight. Till then, let's focus on your case.

**Comment by KaniSchotski** [comment]
Hi Marcin,

Thank you for your quick reply. I did some more testing. And I managed to generate a pdf with images in a table, in a row and inlined (did not try grid yet). 

The problem appears to be related to the size of the images. As soon as the size of the images gets too big, the pdf does no longer get generated. 

It looks like the "row" presentation can handle the biggest images, say images between 100 kb en 150kb per image.  "Table" and "Inlined" can handle images between 50kb and 75 kb per image. 

I will do some more testing. Thusfar, I only tested with 6 images on a line. I will try to generate pdf's with 100, 200, ... images and see where it gets me. 

Regards,
Kani

**Comment by KaniSchotski** [comment]
using QuestPDF.Drawing;
using QuestPDF.Fluent;
using QuestPDF.Infrastructure;
using QuestPDF.Helpers;
using System.Diagnostics.Contracts;

namespace TestPdf.Client.Repos
{
    public class ImageDocument : IDocument
    {
        //the image will appear 6 times per row
        public byte[] Image { get; }
        //the number of rows that will be generated
        public int Rows { get; }

        public ImageDocument(byte[] image, int rows)
        {
            Image = image;
            Rows = rows;
        }

        public DocumentMetadata GetMetadata() => DocumentMetadata.Default;

        public void Compose(IDocumentContainer container)
        {
            container
                .Page(page =>
                {
                    page.Margin(15);
                    page.Size(PageSizes.A4.Landscape());
                    page.Content().Element(ComposeContent);
                });

        }

        void ComposeContent(IContainer container)
        {
            container.PaddingVertical(40).Column(column =>
            {
                column.Spacing(5);
                for (int i = 0; i < Rows; i++)
                {
                    column.Item().Element(ComposeRow);
                }
            });
        }

        void ComposeRow(IContainer container)
        {
            container.Row(row =>
            {
                //row.Spacing(2);

                for (int i = 0; i < 6; i++)
                {
                    row.RelativeItem().Image(Image, ImageScaling.FitArea);
                }
            });
        }
    }
}

**Comment by KaniSchotski** [comment]
Based on the repository above, I did some testing in Blazor. 

The inputs are : 
- an image of type byte[]
- the number of rows of type integer

Via testing, I checked how many rows -  with 6 images per row - Questpdf could handle. An example of 2 rows with an image :

![image](https://user-images.githubusercontent.com/94176942/211166671-190d7942-556b-4d06-b279-d8c3a74c0d01.png)

I tested various images with different pixels, dpi, sizes. Below the results :

![image](https://user-images.githubusercontent.com/94176942/211166748-e35f6db0-d1c7-44ea-a33c-65ac4e559add.png)

Conclusion : regardless of dpi or pixels, the number of bytes of the image determines how many images QuestPdf can handle. The file size of the generated PDF seems to have a maximum of 1,5 Mb. The number of images handled can be estimated by dividing the 1,5 Mb by the size of the image.

It would be nice to be able to set a parameter to increase the maximum resulting file size of the PDF. Be it .5 Mb or 152 Mb.

Please pass me your advice Marcin.

Regards,
Kani
